### PR TITLE
consensus/clique: implement getSignerForBlock

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
Prints the signer for a certain block

### CODE REVIEW

**Comment by MariusVanDerWijden** [comment]
@holiman suggested to have `getSealerForBlockRlp`

**Comment by karalabe** [comment]
Could we do `getSigner` (i.e. no block) and allow specifying either a number a hash (these 2 are easy, we have a rpc.hashOrNumber type) and also perhaps an RLP hex dump (block will always be >> than 32byte hash, so no clash). I think it would be useful to multiplex the various calls one might make without having to remember 3 method names. Just my 2c.

File: consensus/clique/api.go [hunk 1]
```diff
@@ -175,3 +176,21 @@ func (api *API) Status() (*status, error) {
 		NumBlocks:     numBlocks,
 	}, nil
 }
+
+func (api *API) GetSignerForBlock(number *rpc.BlockNumber) (common.Address, error) {
```
Comment (karalabe) [inline consensus/clique/api.go:180]: Doxx it

**Comment by MariusVanDerWijden** [comment]
So https://github.com/ethereum/go-ethereum/pull/22987/commits/890673abf055b8943ca02c116d5d64a7751d1ab8 implements `getSignerForBlock` (doxxed and pretty decently)
While https://github.com/ethereum/go-ethereum/pull/22987/commits/168c82dcaceaf69515fb9250f0339af83f7d0f8c implements `getSigner`in a way that lets us take the RLP too. I haven't tested the second one yet. I believe that it is not working yet due to the `inputBlockNumberFormatter`in web3.js

File: consensus/clique/api.go [hunk 2]
```diff
@@ -175,3 +179,54 @@ func (api *API) Status() (*status, error) {
...
+		if blockNrOrHash == nil {
+			header = api.chain.CurrentHeader()
+		} else if hash, ok := blockNrOrHash.Hash(); ok {
+			header = api.chain.GetHeaderByHash(hash)
+		} else if number, ok := blockNrOrHash.Number(); ok {
+			header = api.chain.GetHeaderByNumber(uint64(number.Int64()))
+		}
+		return getSigner(header)
+	}
+	blob := common.Hex2Bytes(rlpOrBlockNr.RLP)
```
Comment (holiman) [inline consensus/clique/api.go:204]: Might be nice to use a method which gives back an error on invalid hex input

File: internal/web3ext/web3ext.go [hunk 3]
```diff
@@ -108,6 +108,12 @@ web3._extend({
 			call: 'clique_status',
 			params: 0
 		}),
+		new web3._extend.Method({
+			name: 'getSigner',
+			call: 'clique_getSigner',
+			params: 1,
+			inputFormatter: [web3._extend.formatters.inputBlockNumberFormatter]
```
Comment (holiman) [inline internal/web3ext/web3ext.go:115]:
```suggestion
			inputFormatter: [null]
```
Otherwise it will surely reject rlp hexdata, no?

**Review by holiman — COMMENTED** [review]
Looks like a really useful api endpoint!

File: internal/web3ext/web3ext.go (hunk 3 above)
Comment (MariusVanDerWijden) [reply internal/web3ext/web3ext.go]: Weirdly it did not reject rlp hexdata, but I removed it anyway

File: consensus/clique/api.go [hunk 4]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
 		NumBlocks:     numBlocks,
 	}, nil
 }
+
+type SignerBlock struct {
```
Comment (MariusVanDerWijden) [inline consensus/clique/api.go:186]: blockRLPorHash

File: consensus/clique/api.go [hunk 5]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+		if blockNrOrHash == nil {
+			header = api.chain.CurrentHeader()
+		} else if hash, ok := blockNrOrHash.Hash(); ok {
+			header = api.chain.GetHeaderByHash(hash)
+		} else if number, ok := blockNrOrHash.Number(); ok {
+			header = api.chain.GetHeaderByNumber(uint64(number.Int64()))
+		}
+		return getSigner(header)
+	}
+	blob, err := hex.DecodeString(strings.TrimLeft(rlpOrBlockNr.RLP, "0x"))
```
Comment (karalabe) [inline consensus/clique/api.go:224]: You probably want to use `hexutil`

File: consensus/clique/api.go [hunk 6]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
 		NumBlocks:     numBlocks,
 	}, nil
 }
+
+type SignerBlock struct {
+	*rpc.BlockNumberOrHash
+	RLP string `json:"rlp,omitempty"`
```
Comment (karalabe) [inline consensus/clique/api.go:188]: We should use `hexutil.Bytes`, which solves a whole lot of problems.

File: consensus/clique/api.go [hunk 7]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+			header = api.chain.GetHeaderByNumber(uint64(number.Int64()))
+		}
+		return getSigner(header)
+	}
+	blob, err := hex.DecodeString(strings.TrimLeft(rlpOrBlockNr.RLP, "0x"))
+	if err != nil {
+		return common.Address{}, err
+	}
+	block := new(types.Block)
+	if err := rlp.Decode(bytes.NewReader(blob), block); err == nil {
```
Comment (karalabe) [inline consensus/clique/api.go:229]: rlp.DecodeBytes

File: consensus/clique/api.go [hunk 8]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+		return getSigner(block.Header())
+	}
+	header := new(types.Header)
+	if err := rlp.Decode(bytes.NewReader(blob), header); err == nil {
```
Comment (karalabe) [inline consensus/clique/api.go:233]: rlp.DecodeBytes

File: consensus/clique/api.go [hunk 9]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+		return getSigner(header)
+	}
+	return common.Address{}, fmt.Errorf("could not decode rlp")
+}
+
+func getSigner(header *types.Header) (common.Address, error) {
+	if header == nil {
+		return common.Address{}, errors.New("could not find header")
+	}
```
Comment (karalabe) [inline consensus/clique/api.go:242]: Lets not add defensive checks for non API (untrusted) methods.

File: consensus/clique/api.go [hunk 10]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+		return getSigner(block.Header())
+	}
+	header := new(types.Header)
+	if err := rlp.Decode(bytes.NewReader(blob), header); err == nil {
+		return getSigner(header)
+	}
+	return common.Address{}, fmt.Errorf("could not decode rlp")
+}
+
+func getSigner(header *types.Header) (common.Address, error) {
```
Comment (karalabe) [inline consensus/clique/api.go:239]: getSigner == clique.Author

File: consensus/clique/api.go [hunk 11]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+	if err != nil {
+		return common.Address{}, err
+	}
+	block := new(types.Block)
+	if err := rlp.Decode(bytes.NewReader(blob), block); err == nil {
+		return getSigner(block.Header())
+	}
+	header := new(types.Header)
+	if err := rlp.Decode(bytes.NewReader(blob), header); err == nil {
+		return getSigner(header)
```
Comment (karalabe) [inline consensus/clique/api.go:234]: `return api.clique.Author(header)`

File: consensus/clique/api.go [hunk 12]
```diff
@@ -175,3 +182,74 @@ func (api *API) Status() (*status, error) {
...
+func (sb *SignerBlock) UnmarshalJSON(data []byte) error {
+	bnOrHash := new(rpc.BlockNumberOrHash)
+	// Try to unmarshal bNrOrHash
+	if err := bnOrHash.UnmarshalJSON(data); err == nil {
+		sb.BlockNumberOrHash = bnOrHash
+		return nil
+	}
+	// Try to unmarshal RLP
+	var input string
+	err := json.Unmarshal(data, &input)
```
Comment (karalabe) [inline consensus/clique/api.go:200]: You should put these into the same line :D

File: consensus/clique/api.go [hunk 13]
```diff
@@ -175,3 +178,51 @@ func (api *API) Status() (*status, error) {
...
+	}
+	block := new(types.Block)
+	if err := rlp.DecodeBytes(rlpOrBlockNr.RLP, block); err == nil {
+		return api.clique.Author(block.Header())
+	}
+	header := new(types.Header)
+	if err := rlp.DecodeBytes(rlpOrBlockNr.RLP, header); err == nil {
+		return api.clique.Author(header)
+	}
+	return common.Address{}, fmt.Errorf("could not decode rlp")
```
Comment (karalabe) [inline consensus/clique/api.go:227]: This seems weird. IMHO you should do

```go
if err := rlp.DecodeBytes(rlpOrBlockNr.RLP, header); err != nil {
		return common.Address{}, err
	} 
return api.clique.Author(header)
```

**Review by karalabe — APPROVED** [review]
SGTM

**Pull request merged**